borsh/src/de/mod.rs

```diff
@@ -143,6 +143,73 @@ impl BorshDeserialize for usize {
     }
 }
 
+mod non_zero {
+    use super::*;
+    use std::num::*;
+
+    trait NonZero {
+        type Primitive;
+    }
+
+    macro_rules! non_zero_primitives {
+        ($($type:ty => $primitive:ty,)+) => {
+            $(
+                impl NonZero for $type {
+                    type Primitive = $primitive;
+                }
+            )+
+        };
+    }
+
+    non_zero_primitives! {
+        NonZeroU8   => u8,
+        NonZeroU16  => u16,
+        NonZeroU32  => u32,
+        NonZeroU64  => u64,
+        NonZeroU128 => u128,
+
+        NonZeroI8   => i8,
+        NonZeroI16  => i16,
+        NonZeroI32  => i32,
+        NonZeroI64  => i64,
+        NonZeroI128 => i128,
+
+        NonZeroUsize => usize,
+        NonZeroIsize => isize,
```

Comment:
If I am reading code correctly, for serializing `92usize` this'll use 8 bytes on 64 bit system and 4 bytes on 32 bit system. This is wrong, serialization format should be independent of the host architecture. See how the plain `usize` is handled 

https://github.com/near/borsh-rs/blob/735686d9345202e8b8ee1f35eab52d7db6bb82c5/borsh/src/ser/mod.rs#L76-L81

To avoid this (and potentially other) gotchas, I suggest to change the implementation to *delegate* to impls of primiteve types, rather than re-implement the logic. That is, BorshDeserialize for NonZeroU32 should  call `<u32 as BorshDeserialize>::desrialize`, and then do just a single additional check for zero.

Reply:
Done, probably